EKF/control.cpp

```diff
@@ -757,10 +757,19 @@ void Ekf::controlHeightFusion()
 		// we have just switched to using range finder, calculate height sensor offset such that current
 		// measurment matches our current height estimate
 		if (_control_status_prev.flags.rng_hgt != _control_status.flags.rng_hgt) {
-			if (_terrain_initialised) {
-				_hgt_sensor_offset = _terrain_vpos;
+			// use the parameter rng_gnd_clearance if on ground to avoid a noisy offset initialization (e.g. sonar)
+			if (_control_status.flags.in_air) {
+
+				if (_terrain_initialised) {
+
+					_hgt_sensor_offset = _terrain_vpos;
+				} else {
+
+					_hgt_sensor_offset = _R_rng_to_earth_2_2 * _range_sample_delayed.rng + _state.pos(2);
```

Comment:
Yes exactly, I would like the estimate to converge to 0.5m in the above example. Range sensor directly measures height, so I think the offset should always be 0 (or a fixed value defined by where the sensor is mounted on the drone)  when range sensor is primary.
I understand that offsets are required to fuse everything without trouble. So yes maybe a z estimate reset + update of baro offset would work.